plugins/chat/app/services/chat/create_direct_message_channel.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+module Chat
+  # Service responsible for creating a new direct message chat channel.
+  # The guardian passed in is the "acting user" when creating the channel
+  # and deciding whether the actor can communicate with the users that
+  # are passed in.
+  #
+  # @example
+  #  Service::Chat::CreateDirectMessageChannel.call(
+  #    guardian: guardian,
+  #    target_usernames: ["bob", "alice"]
+  #  )
+  #
+  class CreateDirectMessageChannel
+    include Service::Base
+
+    # @!method call(guardian:, **params_to_create)
+    #   @param [Guardian] guardian
+    #   @param [Hash] params_to_create
+    #   @option params_to_create [Array<String>] target_usernames
+    #   @return [Service::Base::Context]
+
+    policy :can_create_direct_message
+    contract
+    model :target_users
+    policy :does_not_exceed_max_direct_message_users,
+           class_name: Chat::DirectMessageChannel::MaxUsersExcessPolicy
+    model :user_comm_screener
+    policy :acting_user_not_disallowing_all_messages
+    policy :acting_user_can_communicate_with_all_parties,
+           class_name: Chat::DirectMessageChannel::CanCommunicateAllPartiesPolicy
+    model :direct_message, :fetch_or_create_direct_message
+    model :channel, :fetch_or_create_channel
+    step :update_memberships
+    step :publish_channel
+
+    # @!visibility private
+    class Contract
+      attribute :target_usernames
+
+      before_validation do
+        self.target_usernames =
+          (
+            if target_usernames.is_a?(String)
```

Comment:
I think it makes sense, since otherwise we would have to introduce logic back into the controller level (which is calling the service) to do the same thing. IMO better to just take whatever `params[:target_usernames]` is and sort it out in the contract before validation.

Reply:
> I think it makes sense, since otherwise we would have to introduce logic back into the controller level (which is calling the service) to do the same thing. IMO better to just take whatever `params[:target_usernames]` is and sort it out in the contract before validation.

Not sure I follow. Can't you just handle differently the `target_usernames` param inside the service based on its type?

Reply:
This is what the contract does at the moment, no? If it’s a string, split it into an array, otherwise don’t do anything.

Reply:
Sorry that wasn't clear, I meant we could have two params maybe?

Reply:
By the way, I’ve found a way of simplifying the contract a bit:
```ruby
    class Contract
      attribute :target_usernames

      before_validation { self.target_usernames = target_usernames.split(",").flatten }

      validates :target_usernames, presence: true
    end
```
`#split` is defined both on `Array` and `String` so we can just call it without having to perform a check first. Then `#flatten` is useful because `Array#split` will return an array of arrays.

I’ve removed `length: { minimum: 1 }` because we’re already validating presence, so the validation will pass only if there is at least one element.

Reply:
> Sorry that wasn't clear, I meant we could have two params maybe?

I think it would complexify things instead. Ideally, if we could do something like `attribute :target_usernames, :array` with auto-casting, it would be perfect :grin:

Reply:
Thanks @Flink will go with the simplified contract. I too think adding 2 params will be needlessly complex when we can just accept both in 1 param and fix it in the contract.

Reply:
I’ve just created this small PR: https://github.com/discourse/discourse/pull/22278.
Having a simple way of handling arrays is always nice, so tell me what you think :grin: @jjaffeux @martin-brennan

Reply:
Thanks @Flink I merged that PR and will use it here :+1: